Fixed deadlocks, completion blocks async, race condition on cancel

bpoplauschi opened the pull request:
#### Fixed issues #625 #507 aka deadlocks (replaces #509)
- In order to fix the deadlock, reviewed the `[SDImageCache diskImageExistsWithKey:]` method. Based on the Apple doc for `NSFileManager`, using the `defaultManager` without the dispatch on the `ioQueue` to avoid the deadlocks. This instance is thread safe. Also created an async variant of this method `[SDImageCache diskImageExistsWithKey:completion:]`
For consistency, added async methods in `SDWebImageManager` `cachedImageExistsForURL:completion:` and `diskImageExistsForURL:completion:`
-  More fixes against deadlocks. `dispatch_sync` on a serial queue is an [anti pattern](http://www.raywenderlich.com/60749/grand-central-dispatch-in-depth-part-1), as it can lead to deadlocks. I replaced all the `dispatch_main_sync_safe` with `dispatch_main_async_safe`, based on the fact that the completion blocks are not returning anything, so we don't need to wait for them to finish.
- The biggest change (architectural) is that the completion block will no longer be executed inside the operation. But that is not an issue, as `NSOperation` does the same (completion block gets called after operation `isFinished`).

#### Fixed race condition In SDWebImageManager (replaces #699)
- In SDWebImageManager if one operation is cancelled, the completion block must not be called, otherwise it might race with a newer completion for the same object

croath commented:
Hey guys I don't know if you guys are still focusing on the issue. Actually I'm not involving with any issues here, just feeling curious about the `sync` things around the callback block, well, while reading the source code of the project.

After reading the explanations from @matej and @rs , and the great example about `-[SDWebImageManager downloadImageWithURL:options:progress:completed:]` hitting the memory cache, I kind of know the reason of using `sync`.

But talking about the UI redrawing, is the main thread just here waiting for memory cache returns back and will draw the image once the image comes back from memory, immediately? There are still lot of time and chances for UI to redraw between querying cache and sending back the `completionBlock`.

And what if the `-[SDWebImageManager downloadImageWithURL:options:progress:completed:]` call couldn't hit any cache and then just fetches it from internet? After the image being downloaded successfully, should we just use a `async` instead of `sync` to call the `completionBlock`?

Please let me know if I missed something or there was any misunderstanding.

Thanks!

bpoplauschi commented:
Redone this via commits 062e50a f7e8246 c77adf4 265ace4 0c47bc3 on the 4.x branch. Same stuff: replaced all sync calls to the completion with async, cleaned up the code a bit. 

Another important change is based on http://blog.benjamin-encz.de/post/main-queue-vs-main-thread. This article states that the main thread may execute code from other queues than the main queue, so we actually want to check for main queue to assure our UIKit operations.
We used: `strcmp(dispatch_queue_get_label(DISPATCH_CURRENT_QUEUE_LABEL), dispatch_queue_get_label(dispatch_get_main_queue())`

bpoplauschi commented:
@croath I see your point, but I feel safer with the async code, since we have a history of locking here and there. Take a look and let me know what you think.

croath commented:
@bpoplauschi Awesome! Yep the safety, just what I was talking about. So that I asked about why `sync` but not `async`. Maybe we can see the result differs with profiling work in some specific situations.
